### ISSUE TITLE
Decouple balances module with other modules

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#1514]
Currently following srml modules depends on the balances module:
- contract
- council
- democracy
- staking
- treasury

This make it hard to replace/modify balance module with something else (e.g. multi token assets module or UTXO module). Because it means each of the dependent modules needs to be modified.

It will be good to refactor balances module so that people can create a runtime with all the above modules, a different implementation of the balance/token/currency module, and have them work together.

I see this could be complicated to make it generalized enough to support all the possible token models so a good start point could be remove hard dependency between balance module and other modules
  - Abstract the balance module interface
  - Modify other modules to reference the balance module interface (via generic) instead of depends on the balance module directly
  - Update the runtime construction code to supply the balances module as the concrete implementation of the balance interface
- As the result, it is possible to provide an alternative balances module (e.g. UTXO) without modify other modules
- This should allow each module reference to different balances implementation. e.g. Staking use existing balances module whereas contract use the assets module.

Basically similar to how executive module use `MakePayment` trait to avoid linking to balances module directly.

I see three traits can be used to replace dependency of balances module
- MakePayment
  - Tx fee payment, use by executive and contract
- TransferToken
  - transfer token between accounts, use by contract, treasury
- Staking
  - reserve, slash, reward, use by staking, council, democracy, treasury

**Comment by gavofyork** [comment]
We may want to get this into the 1.0 release.

**Comment by thiolliere** [comment]
So decouple is made for council, democracy, staking, treasury.

Remains contract to do

**Comment by thiolliere** [comment]
related: https://github.com/paritytech/substrate/pull/1815

**Comment by thiolliere** [comment]
I'll unassign myself for the moment
